### PR TITLE
SONARJAVA-1295 Support proper exit path in try finally

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 1]
```diff
@@ -235,8 +234,14 @@ private void handleBlockExit(ExplodedGraph.ProgramPoint programPosition) {
       }
     }
     // unconditional jumps, for-statement, switch-statement, synchronized:
-    for (CFG.Block successor : block.successors()) {
-      enqueue(new ExplodedGraph.ProgramPoint(successor, 0), programState);
+    if(node.exitPath) {
+      enqueue(new ExplodedGraph.ProgramPoint(block.exitBlock(), 0), programState, true);
+    } else {
+      for (CFG.Block successor : block.successors()) {
+        if (!block.isFinallyBlock() || successor != block.exitBlock()) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

File: java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 2]
```diff
@@ -235,8 +234,14 @@ private void handleBlockExit(ExplodedGraph.ProgramPoint programPosition) {
...
+          enqueue(new ExplodedGraph.ProgramPoint(successor, 0), programState, successor == block.exitBlock());
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

File: java-squid/src/main/java/org/sonar/java/cfg/CFG.java [hunk 3]
```diff
@@ -803,7 +834,11 @@ private Block createUnconditionalJump(Tree terminator, @Nullable Block target) {
     Block result = createBlock();
     result.terminator = terminator;
     if (target != null) {
-      result.addSuccessor(target);
+      if (target == exitBlock()) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/cfg/CFG.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

File: java-squid/src/main/java/org/sonar/java/se/ExplodedGraph.java [hunk 4]
```diff
@@ -84,6 +84,7 @@ public boolean equals(Object obj) {
     final ProgramPoint programPoint;
     @Nullable
     final ProgramState programState;
+    public boolean exitPath = false;
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/ExplodedGraph.java]: MAJOR Make exitPath a static final constant or non-public and provide accessors if needed. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.

File: java-squid/src/main/java/org/sonar/java/se/ExplodedGraph.java (hunk 4 above)
Comment (DidierBesset) [inline java-squid/src/main/java/org/sonar/java/se/ExplodedGraph.java]: I think this is the way to go: the program state should memorize the mode of executing a finally block (with successors or exit path)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.

File: java-squid/src/test/files/se/NullDereferenceCheck.java [hunk 5]
```diff
@@ -1,6 +1,7 @@
 @SomeAnnotation(name = value)
 package javax.annotation;
 
+import java.lang.Exception;
```
Comment (mpaladin) [inline java-squid/src/test/files/se/NullDereferenceCheck.java]: @benzonico not needed.

File: java-squid/src/test/files/se/Reproducer.java [hunk 6]
```diff
@@ -1,5 +1,7 @@
 package javax.annotation;
 
+import java.lang.Exception;
```
Comment (mpaladin) [inline java-squid/src/test/files/se/Reproducer.java]: @benzonico not needed.

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.

**Pull request merged**